source/common/network/connection_impl.cc

```diff
@@ -324,9 +324,9 @@ void ConnectionImpl::write(Buffer::Instance& data, bool end_stream) {
     // TODO(mattklein123): All data currently gets moved from the source buffer to the write buffer.
     // This can lead to inefficient behavior if writing a bunch of small chunks. In this case, it
     // would likely be more efficient to copy data below a certain size. VERY IMPORTANT: If this is
-    // ever changed, read the comment in Ssl::ConnectionImpl::doWriteToSocket() VERY carefully.
-    // That code assumes that we never change existing write_buffer_ chain elements between calls
-    // to SSL_write(). That code will have to change if we ever copy here.
+    // ever changed, read the comment in SslSocket::doWrite() VERY carefully. That code assumes that
```

Comment:
@mattklein123 does this part still apply? With this PR `SslSocket::doWrite` doesn't depend on each element but instead linearize first `bytes_to_write` bytes and pass it to `SSL_write`. So the code doesn't assume "we never change existing write_buffer_ chain elements between calls to SSL_write()", just assuming we never drain data from the buffer between calls to SSL_write(), no?

Reply:
You can append, but you cannot drain, which is the same as before changes in this PR, AFAIK.

Reply:
IIUC before this PR, a buffer `{"abc"}`, couldn't be `{"abcde"}` in this method, you can only add one or more slice then make it into `{"abc","de"}`. With this PR, either `{"abcde"}` or `{"abc","de"}` will work since you captures `bytes_to_write` in SslSocket::doWrite.

Reply:
Per @davidben, you can pass more data to the retry call (but it must start with the same data as the failed call), so `{"abc"}` -> `{"abcde"}` was also fine before this PR.

Capturing `bytes_to_retry_` is really an optimization to avoid linearizing over data that's going to be sent over multiple TLS records that doesn't need to be combined together.

Reply:
Yeah after reading @davidben's comment I noticed that `{"abc"} -> {"abcde"}` was also find before this PR, but I think the original intention is not to allow that case, no? @mattklein123

Reply:
I think when I wrote this code I was under the impression that the parameters had to be exactly the same (pointer and length). If that's not the case, then the comment is just wrong. @PiotrSikora can you fix my comment or just remove it if no longer relevant? (Does the pointer still have to be the same though? If so we should keep that part of the comment).

Reply:
The pointer needs to be the same by default but it's just a silly sanity check. If your buffer may be reallocated or whatever, set the moving buffer option.

Reply:
The way to think about it is it's like POSIX write where you either get told the write failed with a retryable error or it finished completely with N bytes written.

The difference is that write allows this:
```
write("hello") => EWOULDBLOCK
write("actually I changed my mind")
```
with the first input never making it on the wire. This shape of API doesn't compose well with TLS if we managed to write half a record. (Writing half a record does not mean you wrote half the plaintext. It means you half-wrote the full plaintext and can't change your mind.)

OpenSSL's API handles this case by returning EWOULDBLOCK (so it looks to you like nothing was written) and then says you are not allowed to change your mind on what you are writing. Once you pass input in, you gave committed to that input. It enforces this with some sanity checks on the length and optionally the pointer, but the actual rule is the "no rewinding" rule. You want to think about that one and not the length business.

Reply:
@mattklein123 it doesn't have to be the same pointer and length, but the data needs to be the same, so that comment isn't completely wrong, IMHO. And the only reason why the pointer doesn't need to be the same is because `SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER` is set, but it doesn't look that there is a reason for it, so I'm going to remove it, which will make the comment a bit more accurate.

Reply:
SGTM, as long as you checked the comment and you think it's good it's fine with me. Thank you all for the very lively and enlightening discussion on this. :)

Reply:
SGTM.